Issue 229: arg check on admincommands

Veritania opened the pull request:
A different approach to solve Issue #229, compared to my previous suggestion in PR #281

### Summary 
This way we validate that all the required types are present.
If the required values are null, or we can't Parse them we ` throw PerpetuumException.Create(ErrorCodes.RequiredArgumentIsNotSpecified)`.

If an exception is not of types `ArgumentNullException` or `FormatException` we throw a generic exception as a safety measure. In reality we should only run into a `ArgumentNullException`.

- Another form of validation we need, is to make sure the values are within certain bounds (specifics to be determined for each command)
- - Example: Say we can parse a `int x` coord from the string args[], how do we know its valid coord and not something ridiculous like 999999

@MikeJeffers Could you confirm if this is an approach you'd want for closing #229 ? Before I go ahead and convert more of the commands into this style.

Veritania committed:
Validate Args and split logic to new class - zoneAddDecor

Veritania committed:
validation and logic in same file

Veritania committed:
arg check on jumpTo

Veritania committed:
shorten code

Veritania committed:
clean space

Veritania committed:
better exception handling

MikeJeffers commented:
Yes, I think this is the basic error handling that needs to be integrated into each of the commands.  
Because if you look at some of the command handlers, the use of tryParse just replaces a shared bool value and only the last is validated, which is broken.

I think if you want to get more sophisticated, maybe with more specific checks relevant to particular commands (like jumpto checks size of the zone, or presence of the zone in the zonemanager), thats fine too, but maybe in a separate PR.

The 3rd level of enhancement (totally out of scope, but would be cool) is almost like a signature for each command method, an ordered dict of args and their types and optional/requirements.  So we could do something like `#cmd, --help` and get the expected args.  Again, not a high priority, anyone using these commands is looking at source anyway, or we document the ones the team uses frequently on the wiki.

Veritania committed:
Arg check on FlagPlayerNameOffensive

Veritania committed:
argcheck EPBonusSet & UnlockAllEP

Veritania committed:
argcheck on SetWeather

Veritania committed:
argcheck on MovePlayer

Veritania committed:
argcheck ZoneCreateGarden

Veritania committed:
argcheck on SpawnRelic

Veritania committed:
argcheck ZoneAddDecorToLockedTile

Veritania committed:
argcheck ZoneLockDecor

Veritania commented:
There's no more scenarios left where we incorrectly reassign a bool variable to error-check multiple arguments and then only do validation on the last entry.

So the main problem is now solved, but there's a case to be made to streamline this entire class to all use the same style in how it assigns and checks for arguments etc (basically what was said in this https://github.com/OpenPerpetuum/PerpetuumServer/pull/291#issuecomment-808825309 )

I think we could open new issues to target these, and close this PR.

MikeJeffers commented:
> 
> 
> There's no more scenarios left where we incorrectly reassign a bool variable to error-check multiple arguments and then only do validation on the last entry.
> 
> So the main problem is now solved, but there's a case to be made to streamline this entire class to all use the same style in how it assigns and checks for arguments etc (basically what was said in this [#291 (comment)](https://github.com/OpenPerpetuum/PerpetuumServer/pull/291#issuecomment-808825309) )
> 
> I think we could open new issues to target these, and close this PR.

I agree, fancy refactor can be a separate PR/issue

MikeJeffers commented on src/Perpetuum/Services/Channels/ChatCommands/AdminCommandHandlers.cs:
```diff
@@ -170,13 +170,21 @@ public static void ShutdownCancel(AdminCommandData data)
-            err = !int.TryParse(data.Command.Args[1], out int x);
-            err = !int.TryParse(data.Command.Args[2], out int y);
-            if (err)
+            int zone;
```
Here is another quick thing to add in as these get refactored:

Anytime a zone parameter is being checked, you can also call the helper:
```
CheckZoneId(data, zoneId);
```
Which will check against the set of active zones, throwing if the id is not for an active/valid zone.

MikeJeffers reviewed (changes requested):
Use those helper functions!

MikeJeffers commented on src/Perpetuum/Services/Channels/ChatCommands/AdminCommandHandlers.cs:
```diff
@@ -170,13 +170,21 @@ public static void ShutdownCancel(AdminCommandData data)
-                throw PerpetuumException.Create(ErrorCodes.RequiredArgumentIsNotSpecified);
+                zone = int.Parse(data.Command.Args[0]);
+                x = int.Parse(data.Command.Args[1]);
+                y = int.Parse(data.Command.Args[2]);
```
Aha! another thing to check first (or catch and report on) is the expected argument length!
For example here, if only the zoneid is passed, or just 2 params (ints) this will have an out of bounds array access, which throws whatever the error is, but isn't informative as to the issue.

That said, it is better than it was before - but another exception type check could handle this nicely

Veritania committed:
Add CheckRequiredArgs method and utilize CheckZoneId

Veritania commented:
I've added a helper method to the class, which checks the given number of args compared to the expected number of args.
It's only being used on the commands that has a set number of required args - The ones which can have optional args I haven't included.

And added the CheckZoneId where it was relevant on the methods I've made changes to.

Veritania committed:
check req length + cleanup

MikeJeffers commented on src/Perpetuum/Services/Channels/ChatCommands/AdminCommandHandlers.cs:
```diff
@@ -106,6 +106,15 @@ public static void CheckZoneId(AdminCommandData data, int zoneId)
+            if(data.Command.Args.Length != expectedLength)
+            {
+                SendMessageToAll(data, $"Error - Command requires " + expectedLength.ToString() + " args, but was given " +data.Command.Args.Length.ToString()+ " args");
+                throw PerpetuumException.Create(ErrorCodes.TooManyOrTooFewArguments);
```
VERY cool - like this a lot

MikeJeffers commented on src/Perpetuum/Services/Channels/ChatCommands/AdminCommandHandlers.cs:
```diff
@@ -256,8 +273,10 @@ public static void MovePlayer(AdminCommandData data)
+
             int.TryParse(data.Command.Args[0], out int definition);
-            int.TryParse(data.Command.Args[1], out int qty);
+            int.TryParse(data.Command.Args[1], out int qty);   
```
trailing whitespace.. eww

MikeJeffers reviewed (approved):
Sorry that last one should have been approve.  Style stuff can be fixed before merge.

Veritania committed:
eww whitespace fix
